server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -277,61 +276,101 @@ private boolean skipRepository(String repositoryName) {
             }
         }
 
-        private void loadSnapshotInfos(String repo, @Nullable RepositoryData repositoryData, ActionListener<Void> listener) {
+        private void loadSnapshotInfos(String repositoryName, @Nullable RepositoryData repositoryData, ActionListener<Void> listener) {
             assert ThreadPool.assertCurrentThreadPool(ThreadPool.Names.MANAGEMENT);
 
-            if (cancellableTask.notifyIfCancelled(listener)) {
-                return;
-            }
-
-            final Set<String> unmatchedRequiredNames = new HashSet<>(snapshotNamePredicate.requiredNames());
-            final Set<Snapshot> toResolve = new HashSet<>();
-
-            for (final var snapshotInProgress : snapshotsInProgress.forRepo(repo)) {
-                final var snapshotName = snapshotInProgress.snapshot().getSnapshotId().getName();
-                unmatchedRequiredNames.remove(snapshotName);
-                if (snapshotNamePredicate.test(snapshotName, true)) {
-                    toResolve.add(snapshotInProgress.snapshot());
-                }
-            }
-
-            if (repositoryData != null) {
-                for (final var snapshotId : repositoryData.getSnapshotIds()) {
-                    final var snapshotName = snapshotId.getName();
-                    unmatchedRequiredNames.remove(snapshotName);
-                    if (snapshotNamePredicate.test(snapshotName, false) && matchesPredicates(snapshotId, repositoryData)) {
-                        toResolve.add(new Snapshot(repo, snapshotId));
```

Comment:
I assume `new Snapshot(repo, snapshotId)` should be is largest difference in allocation, and following `toResolve.stream().map(Snapshot::getSnapshotId).toList()`. Am I it right?

Does it make sense to move filtering into `snapshotsInProgress.forRepo` add an extra predicate argument?

Reply:
With this change we no longer allocate `toResolve` (a potentially-enormous `Set`) at all. But then yeah we save the allocation of the (potentially-enormous) list that gets passed to `loadSnapshotInfos`.

I have several further pieces of work here, e.g. to address the fact stuff like the fact that we still allocate the (potentially-enormous) `snapshotIdsToIterate` within `loadSnapshotInfos`. But this seems like a good place to stop for now.